ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,273 @@
+client = Temporal::Client.new(
+  connection,
+  'my-namespace',
+  payload_converter: converter,
+  payload_codecs: codecs
+)
+
+worker = Temporal::Worker.new(
+  connection,
+  namespace: 'demo',
+  task_queue: 'demo',
+  workflows: [...],
+  activities: [...],
+  payload_converter: converter,
+  payload_codecs: codecs
+)
+```
+
+*This does indeed involve some duplication and we might simplify this initialization later
+on. But it's outside the scope for this proposal.*
+
+Notes:
+
+- Converters and codecs both can be classes and class instances as long as the given object
+  implements the required interface
+- This API is still being figured out to accommodate future features and configuration params
+
+
+## Data conversion
+
+All the previously outlined modules are required to facilitate the process of data conversion.
+Here's the diagram representing it:
+
+![Data conversion](./images/data-conversion.png)
+
+This describes the process that every piece of data (workflow/activity args and return values,
+signal inputs, query results, headers, memos, etc) goes through getting transformed from the
+user-land object to something that can be transmitted over the wire.
+
+*NOTE: As mentioned previously, the "Type hints" are optional and will be omitted altogether unless
+defined.*
+
+The reverse process is applied when converting the data from a protobuf received over the wire back
+to a use-land object.
+
+
+## Errors
+
+Here's the Error hierarchy that is planned for the Ruby SDK:
+
+```ruby
+module Temporal
+  # top level error superclass
+  class Error < StandardError
+
+  # superclass for all errors within the SDK itself
+  class InternalError < Error
+
+  class ClientError < InternalError # errors specific to Client
+  class MissingNamespace < ClientError
+  ...
+
+  class WorkerError < InternalError # errors specific to Worker
+  class TooManyStickyWorkflows < WorkerError
+  ...
+
+  # superclass for connection and network errors
+  class NetworkError < Error
+
+  class TimeoutError < NetworkError
+  class HostUnreachable < NetworkError
+  ...
+
+  # superclass for errors related to the server and it's responses
+  class ServerError < Error
+
+  class MissingWorkflowHistory < ServerError
+  class ExpectedSingleCloseEvent < ServerError
+  ...
+
+  # superclass for all API failure responses
+  class APIError < ServerError
+
+  class WorkflowExecutionAlreadyStarted < APIError
+  class NamespaceNotActive < APIError
+  ...
+
+  # superclass for type hinting errors
+  class TypeError < Error
+
+  class WrongType < TypeError
+  class UnsupportedType < TypeError
+  ...
+
+  # superclass for workflow errors
+  class WorkflowError < Error
+
+  class Cancelled < WorkflowError
+  class UnsupportedQuery < WorkflowError
+  class DeadLockDetected < WorkflowError
+  ...
+
+  # superclass for activity errors
+  class ActivityError < Error
+
+  class MissingActivity < ActivityError
+  class MissingHeartbeat < ActivityError
```

Comment:
What are these?

Reply:
`MissingActivity` for when an activity is not registered with a worker and the other one I thought might be used for when an activity fails because of a missed heartbeat (but I guess it manifests itself as a timeout error, so probably not needed)